File: web/src/model/victron_controller/dashboard/v0_2_0/CoresState.ts
// @ts-nocheck
import {BaboonGenerated, BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../../BaboonSharedRuntime'
import {CoreState, CoreState_UEBACodec} from './CoreState'

export class CoresState implements BaboonGenerated {
    private readonly _cores: Array<CoreState>;
    private readonly _topo_order: Array<string>;

    constructor(cores: Array<CoreState>, topo_order: Array<string>) {
        this._cores = cores
        this._topo_order = topo_order
    }

    public get cores(): Array<CoreState> {
        return this._cores;
    }
    public get topo_order(): Array<string> {
        return this._topo_order;
    }

    public toJSON(): Record<string, unknown> {
        return {
            cores: this._cores,
            topo_order: this._topo_order
        };
    }

    public with(overrides: {cores?: Array<CoreState>; topo_order?: Array<string>}): CoresState {
        return new CoresState(
            'cores' in overrides ? overrides.cores! : this._cores,
            'topo_order' in overrides ? overrides.topo_order! : this._topo_order
        );
    }

    public static fromPlain(obj: {cores: Array<CoreState>; topo_order: Array<string>}): CoresState {
        return new CoresState(
            obj.cores,
            obj.topo_order
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return CoresState.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return CoresState.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#CoresState'
    public baboonTypeIdentifier() {
        return CoresState.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.2.0", "0.3.0"]
    public baboonSameInVersions() {
        return CoresState.BaboonSameInVersions
    }
    public static binCodec(): CoresState_UEBACodec {
        return CoresState_UEBACodec.instance
    }
}

/** @deprecated Version 0.2.0 is deprecated, you should migrate to 0.3.0 */
export class CoresState_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: CoresState, writer: BaboonBinWriter): unknown {
        if (this !== CoresState_UEBACodec.lazyInstance.value) {
          return CoresState_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                BinTools.writeI32(buffer, Array.from(value.cores).length);
            for (const item of value.cores) {
                CoreState_UEBACodec.instance.encode(ctx, item, buffer);
            }
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                BinTools.writeI32(buffer, Array.from(value.topo_order).length);
            for (const item of value.topo_order) {
                BinTools.writeString(buffer, item);
            }
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            BinTools.writeI32(writer, Array.from(value.cores).length);
            for (const item of value.cores) {
                CoreState_UEBACodec.instance.encode(ctx, item, writer);
            }
            BinTools.writeI32(writer, Array.from(value.topo_order).length);
            for (const item of value.topo_order) {
                BinTools.writeString(writer, item);
            }
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): CoresState {
        if (this !== CoresState_UEBACodec .lazyInstance.value) {
            return CoresState_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 2; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const cores = Array.from({ length: BinTools.readI32(reader) }, () => CoreState_UEBACodec.instance.decode(ctx, reader));
        const topo_order = Array.from({ length: BinTools.readI32(reader) }, () => BinTools.readString(reader));
        return new CoresState(
            cores,
            topo_order,
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return CoresState_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return CoresState_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#CoresState'
    public baboonTypeIdentifier() {
        return CoresState_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new CoresState_UEBACodec())
    public static get instance(): CoresState_UEBACodec {
        return CoresState_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/v0_2_0/cores_state.rs
use crate::victron_controller::dashboard::v0_2_0::core_state::CoreState;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct CoresState {
    pub cores: Vec<CoreState>,
    pub topo_order: Vec<String>,
}



impl crate::baboon_runtime::BaboonBinCodecIndexed for CoresState {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        2
    }
}

impl crate::baboon_runtime::BaboonBinEncode for CoresState {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                crate::baboon_runtime::bin_tools::write_i32(&mut buffer, value.cores.len() as i32)?;
            for item in (value.cores).iter() {
                item.encode_ueba(ctx, &mut buffer)?;
            }
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                crate::baboon_runtime::bin_tools::write_i32(&mut buffer, value.topo_order.len() as i32)?;
            for item in (value.topo_order).iter() {
                item.encode_ueba(ctx, &mut buffer)?;
            }
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            crate::baboon_runtime::bin_tools::write_i32(writer, value.cores.len() as i32)?;
            for item in (value.cores).iter() {
                item.encode_ueba(ctx, writer)?;
            }
            crate::baboon_runtime::bin_tools::write_i32(writer, value.topo_order.len() as i32)?;
            for item in (value.topo_order).iter() {
                item.encode_ueba(ctx, writer)?;
            }
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for CoresState {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let cores = {
            let count = crate::baboon_runtime::bin_tools::read_i32(reader)? as usize;
            (0..count).map(|_| Ok(CoreState::decode_ueba(ctx, reader)?)).collect::<Result<Vec<_>, Box<dyn std::error::Error>>>()?
        };
        let topo_order = {
            let count = crate::baboon_runtime::bin_tools::read_i32(reader)? as usize;
            (0..count).map(|_| Ok(crate::baboon_runtime::bin_tools::read_string(reader)?)).collect::<Result<Vec<_>, Box<dyn std::error::Error>>>()?
        };
        Ok(CoresState {
            cores,
            topo_order,
        })
    }
}